### ISSUE TITLE
[Status] Backporting bot stuff from bmorel's work

### ISSUE THREAD

**Issue by necessarily-equal** [Unvanquished/Unvanquished#2289]
### Current PR on the subject

(note I'm not including here bmorel's initial PR or the PR that have been closed.)

* [x] https://github.com/Unvanquished/Unvanquished/pull/2100
* [x] https://github.com/Unvanquished/Unvanquished/pull/2102
* [x] https://github.com/Unvanquished/Unvanquished/pull/2184
* [x] https://github.com/Unvanquished/Unvanquished/pull/2185
* [x] https://github.com/Unvanquished/Unvanquished/pull/2254
* [x] https://github.com/Unvanquished/Unvanquished/pull/2259
* [x] https://github.com/Unvanquished/Unvanquished/pull/2285
* [x] https://github.com/Unvanquished/Unvanquished/pull/2286

### Other things I'd like to import

* [ ] bot mind rework with regard to targets and memory
* [ ] " b79285f413 prevent bots to all use the same weapon (esp. luci/rant)" https://github.com/Unvanquished/Unvanquished/issues/1861
* [x] " 9dad14ae3 teach bots how to use luci" https://github.com/Unvanquished/Unvanquished/issues/1663
* [ ] maybe " b922cb1f7 consider teammates for the rush decision" ?
* [x] maaaaaybe the jump thing 1bd08870a and the slopes thing 57e3f4992 if I manage to find an issue in master they would fix
* [x] the crouching code, maybe after adding some weights to the navmesh generator https://github.com/Unvanquished/Unvanquished/issues/1862
* [x] Something which isn't from the betterai branch, but https://github.com/Unvanquished/Unvanquished/tree/EntityUseRange
* [x] https://github.com/Unvanquished/Unvanquished/issues/1865

### Some things I'd like to import but that will need some sort of rewrite
* [x] " 301873946 alien bots no longer dodge fleeing enemies"
* [x] bot going to grab poison from a booster https://github.com/Unvanquished/Unvanquished/issues/1865

### The full commit list with (my opinion of) the state of every commit

<details> 
  <summary>I've hidden this in spoilers, and spoilers-in-spoilers to keep the length of this issue reasonable. Click here to unroll.</summary>

The first commit in the branch is the parent commit in the tree.

Legend:
```
m = commits merged in [m]aster, or cherry-picked from master
d = commits that are redundant and could probably be [d]ropped
l = commits that aren't worth getting into master ([l]ower interest)
p = commits that looks [p]romising and could be [p]icked
e = commits that aren't in master but come from [e]lsewhere (also not to be backported)
z = documentation
b = I don't think backporting this is a good idea
u = there is an open upstream PR for this
```

<details> 
  <summary>Full list of commits</summary>

```
d 3956b3af3 for readability's sake!
d 14bdcbefd code style convertion, to revert after cleanup
l 732c39e59 don't do methods when functions suffice
l 0d1e7287e remove useless functions
m 41b4c530a glm: migrate botlib
m 7893618f1 make botMemory_t::nav private
m 93de7b03f glm: BotGetGoalRadius & RadiusFromBounds2D
m ba0184ae3 prepare botTarget_t
m 6e34364df glm: BotMoveToGoal
m 8913797b6 glm: BotActionAimAtGoal
m 7105b2839 glm: DistanceToGoal2DSquared
m 7799d2ab1 glm: DistanceToGoal
m 059b6c5d2 glm: DistanceToGoalSquared
m ddbd2ad49 glm: botTarget_t::getPos(vec3_t pos) const
m 897f91208 glm: BotClampPos; BG_BoundingBox( class_t ); trap_Trace
m 767f419d9 glm: BotSlowAim internals; BG_GetClientViewOrigin; BG_GetClientNormal
m 963168006 glm: BotSeek
m 875b49781 glm: new functions converted
m 673e1e344 add a BG_BoundingBox for buildables & use it
m 75a6ce336 glm: BotActionHealH (convertion almost complete)
m 38a6d17b1 almost got rid of vec3_t in sg_bot_ai.cpp
m 3d81fb92d glm moves for sg_bot_util.cpp
m 788cbba01 port BotTargetInAttackRange to glm
m 58ec55bd7 almost get rid of vec3_t for sg_bot*
m c88f25d08 glm: botRouteTarget_t::setPos
m 2408934c9 glm: BotAimAngle
m 0585faa18 glm: BotTargetGetBoundingBox internals
m 9ccbd8282 glm: BotTargetInAttackRange
m 440c8c6c4 glm: BotChangeClass; BotEvolveToClass; BotResetStuckTime: minor stuff
m b4be8f440 glm: steal AngleVectors from daemon
m 971fe3ddb glm: ProjectPointOntoVector+minor
m 1a3ea709e glm: simply G_CalcMuzzlePoint API
  2ecb66970 glm: simplify code and move functions in shared
  a8e7dd5dc switched Beacon.cpp to glm
m d50909ac7 switched sg_team.cpp to glm
m d75fbbd74 remove sg_bot unused function
m 8273fad7e switch sg_spawn_gfx.cpp to glm
m 3364ac6f5 switch sg_admin.cpp to glm
m fc41666ff various glm stuff
m ada06b725 sg_bot_util: remove useless param
m 7df8fdc1a glm: remove useless VEC2GLM instances
m c69dd2f77 rmlui: Fix copypasta error in the <numSpawns/> element.
m cf7df3caf medistat: Fix idle unpowered animation.
m bf9f251f9 beacons: Don't allow opening beacon menu as a spec.
m 4e7f1296b graphics: avoid ET_LEV2_ZAP_CHAIN undefined behavior
m ab20bf2d3 dist/configs/config/: fix indent
m 8a7d12b95 rmlui: Fix inconsistent coloring on names after Follow text.
m 69d51d307 Sync submodules
m febdec91f [nobuild] cg_rocket_draw.cpp: mark some class members private
m 393f17848 [nobuild] cg_rocket_draw.cpp: rework LocationElement to be easier on CPU
m 9165300b8 [nobuild] cg_rocket_draw.cpp: rework PositionElement to be easier on CPU
m ff2148416 [buildanew] cg_rocket_draw.cpp: use trailing _ for class members
m c7d87bd1d cg_rocket_draw.cpp: drop speedometer background-color support, as it is unused
m 513c6ac5d Add bot fill votes (#2187)
m 9cb981509 bugfix: command showbans could not be found
m 069475d40 check that bots are allowed to evolve/buy in ::allowed() (#2192)
m 877806e56 negative momentum should not happen
m 4c6aca2ea Add tab-completion for the print_momentum command
m 0be4059a5 Improve offensive advanced dragoon bounce targeting
m 5896edb27 rocket: prevent overflow
m 9847d7e9a Fix spacing in name on chasing screen
m 4928329e5 sgame: Remove a bunch of excess newlines
m 8ff7c9d79 sync submodules
m b68773f4d Rewrite update-version-number in Python
m eaa267f7b sgame: Add g_ignoreNobuild (#2204)
m 64ac76aff Get rid of blueprint miner entity hack
m 7dd792d53 chat: Show a chat bubble when players are typing.
m 147a9cbde Bump Mac CI image to version 11
m c0912cd15 Document ProjectPointOntoVector
m 28eb745f6 Remove unneeded destructor of SpawnerComponent
m 059cb0181 rmlui: Bounds check the chat cursor.
m 0ad72b965 zap: Fix copypaste error in marazap.
m 83922a242 zap: Fix lingering zap due to OB1 error.
m 953b8ddc0 Fix typo in variable name
m edb7e15a1 Document some BT thingy
m eddda4058 Reformat BotPain
m 9aaadee0a 1.f -> 1.0f
m 1829a4ba0 pmove: use one more Math::Clamp
m 7d3dc3d17 rmlui: Properly handle more edge cases in execForm
m be91c065c admin: Fix FIXME where ADMBP_* uses print_tr
m f0beb2ad7 cp: /cp doesn't conflict with any cgame commands
m 24c401c15 Fixes the energy resupply hint, for weapon types and reload places
m 0aa8e0f6b Prepare for const char* from COM_Parse
m 16db33f86 Fix "/team auto" with locked alien team
m b243d9df8 Remove obsolete shell scripts
m 92f15d685 chatbubble: Restrict when chatbubble is down
m b71fafbb4 bg_public.h: replace hex values by bit numbers
m 1979250d4 fix overlapping EF_ flags
e 8c8f3db9b fix MN_WELCOME not working by creating a delay before sending the command to the client. this also specifies that the MN_WELCOME file is called 'changelog' on the client
p caf8b7f54 add HOSTING documentation
p 1dfa02ea2 add PACKAGING (without the "mod" section)
z 76b358806 import a small document about layouts creation
z 27940b2da root folder itself must not be in the archive
z bf35283b8 HOSTING: added more informations
z 79f461a6e doc: add info about nexe file names
m e391dfbee glm: remove no longer implemented function
l 3c61b502a Rework skill system
p b79285f41 prevent bots to all use the same weapon (esp. luci/rant)
l 6d2e4cd9b make bot say where they suicide
l 57e3f4992 fix Slopes of Hell
l 1bd08870a rewrite BotShouldJump
l d42899e49 some cleanup
l 0301cac92 initial work for navcon support
  ded17890a rewrite BotAvoidObstacles
  493c10806 remove old BotAvoidObstacles
  55460a87c implement crouching
  143c00807 fix steering not applied properly
  d2ca3d3f5 checkWalkable doc
  19f136c7a botMemory_t::findPath doc
  7f029fb37 better handling of stepsize
  c2c5e10dd implement hopefully near-final obstacle management
  f948eefe8 disable steer() and detectWalk() since traces do not provide required services
m de73933a8 privateness for stamina handling
  fadded99a bot: aim() != aimSpeed() + no headshots for dumb ones
p 956f1b9a4 implement resupplyScore BT condition
  79bad44e1 implement BT leaves percentAmmoClip, percentClips and reload
  f225be5ab implement BT action BotActionResupply
l f56b97ecb add weaponIsReady BT condition
  e65dc91e1 implement poison support for resupply
m bcd049a43 less branching in pathfinding
m 62e96da0a one-line please
  7a2dd5ceb botMemory_t reorganisation
b 9e633e805 BotChangeGoalEntity -> BotChangeGoal
b ffe50ef15 botTarget_t::operator==
  ffbb13097 complete half-baked GentityRef_impl and botTarget_t
  9ea0f0c7f less botTarget_t target
l 989e3eaa0 BotGetEnemyPriority takes a GentityConstRef
u b04373ffc give botTarget_t::botTarget_t meaningful infos
  5295d4d5f more proposedTarget
  5d4bce5f8 botTarget_t for BotEntityIsValidTarget and BotEntityIsValidEnemyTarget
  9f11d76b7 Rewrite BotSearchForEnemy to use a std::vector
  d69c6db11 pef fixes
m c8b062e66 some BotAimAtEnemy cleanup
  64b2d7ac2 hopefully final fix for enemy memory
u 52abad246 BotDirectionToUsercmd minor cleanup
  53529d535 more weight for current fight
  b922cb1f7 consider teammates for the rush decision
p 9dad14ae3 teach bots how to use luci
u 301873946 alien bots no longer dodge fleeing enemies
  8b3e277bf merge as many .*Attributes_t::stuff as possible into higher level structs (inherit)
  270c7c86f prepare huge .*_RANGE refactoring
  fea37c873 remove ABUILDER_CLAW_RANGE
  e37edeedb remove LEVEL0_BITE_RANGE
  91f0ce02b remove LEVEL1_CLAW_RANGE && LEVEL1_CLAW_U_RANGE
  d92e06ff0 remove LEVEL2_CLAW_RANGE && LEVEL2_U_CLAW_RANGE
  e4d80379b remove PAINSAW_RANGE
  c785f808c remove LEVEL3_CLAW_RANGE && LEVEL3_CLAW_UPG_RANGE
  5d3878d6a remove LEVEL4_CLAW_RANGE
  d8171cd1a small cleanup
  8f7375e26 remove per-melee-weapon width/height/damage variables
  db4ba4963 remove dretch specific variables in favor of weaponAttributes_t
  f72400a1c move painsaw and mantis specifics into FireMelee
  5c15af9b5 remove most shotgun specific variables
  ce5a50649 remove SMG (RIFLE) specific vars
  0ae17c574 remove CHAINGUN specific vars
  81638c4fd remove MDRIVER specific vars
  094070910 remove MDRIVER specific vars
p 760f08c11 get rid of remaining BG_[Class|Buildable]BoundingBox calls
p 80088f668 glm: BuildablesIntersect
m b83410436 remove zombie code
m 5449ebd4f remove trap_CM_BoxTrace (ease of grep)
```

</details>

<details> 
  <summary>The list without m|e|d|u|b commits, that is the list of commits that may be backported</summary>

```
  2ecb66970 glm: simplify code and move functions in shared
  a8e7dd5dc switched Beacon.cpp to glm
p caf8b7f54 add HOSTING documentation
p 1dfa02ea2 add PACKAGING (without the "mod" section)
z 76b358806 import a small document about layouts creation
z 27940b2da root folder itself must not be in the archive
z bf35283b8 HOSTING: added more informations
z 79f461a6e doc: add info about nexe file names
p b79285f41 prevent bots to all use the same weapon (esp. luci/rant)
  ded17890a rewrite BotAvoidObstacles
  493c10806 remove old BotAvoidObstacles
  55460a87c implement crouching
  143c00807 fix steering not applied properly
  d2ca3d3f5 checkWalkable doc
  19f136c7a botMemory_t::findPath doc
  7f029fb37 better handling of stepsize
  c2c5e10dd implement hopefully near-final obstacle management
  f948eefe8 disable steer() and detectWalk() since traces do not provide required services
  fadded99a bot: aim() != aimSpeed() + no headshots for dumb ones
p 956f1b9a4 implement resupplyScore BT condition
  79bad44e1 implement BT leaves percentAmmoClip, percentClips and reload
  f225be5ab implement BT action BotActionResupply
  e65dc91e1 implement poison support for resupply
  7a2dd5ceb botMemory_t reorganisation
  ffbb13097 complete half-baked GentityRef_impl and botTarget_t
  9ea0f0c7f less botTarget_t target
  5295d4d5f more proposedTarget
  5d4bce5f8 botTarget_t for BotEntityIsValidTarget and BotEntityIsValidEnemyTarget
  9f11d76b7 Rewrite BotSearchForEnemy to use a std::vector
  d69c6db11 pef fixes
  64b2d7ac2 hopefully final fix for enemy memory
  53529d535 more weight for current fight
  b922cb1f7 consider teammates for the rush decision
p 9dad14ae3 teach bots how to use luci
  8b3e277bf merge as many .*Attributes_t::stuff as possible into higher level structs (inherit)
  270c7c86f prepare huge .*_RANGE refactoring
  fea37c873 remove ABUILDER_CLAW_RANGE
  e37edeedb remove LEVEL0_BITE_RANGE
  91f0ce02b remove LEVEL1_CLAW_RANGE && LEVEL1_CLAW_U_RANGE
  d92e06ff0 remove LEVEL2_CLAW_RANGE && LEVEL2_U_CLAW_RANGE
  e4d80379b remove PAINSAW_RANGE
  c785f808c remove LEVEL3_CLAW_RANGE && LEVEL3_CLAW_UPG_RANGE
  5d3878d6a remove LEVEL4_CLAW_RANGE
  d8171cd1a small cleanup
  8f7375e26 remove per-melee-weapon width/height/damage variables
  db4ba4963 remove dretch specific variables in favor of weaponAttributes_t
  f72400a1c move painsaw and mantis specifics into FireMelee
  5c15af9b5 remove most shotgun specific variables
  ce5a50649 remove SMG (RIFLE) specific vars
  0ae17c574 remove CHAINGUN specific vars
  81638c4fd remove MDRIVER specific vars
  094070910 remove MDRIVER specific vars
p 760f08c11 get rid of remaining BG_[Class|Buildable]BoundingBox calls
p 80088f668 glm: BuildablesIntersect
```

</details>

Around the end of the branch, the tree is unfortunately somewhat messy where features are sometimes split between several commits and depending on previous commits. Beside what I've listed at the beginning of the issue, I don't think the other commits are worth extracting and importing.

</details>

**Comment by sweet235** [comment]
About " https://github.com/Unvanquished/Unvanquished/commit/b922cb1f76f2fd3e02d3527f1f9eff5f66bc9741 consider teammates for the rush decision":

I tested some bot variants with bmorel, it was great fun. I am not sure if I tested this exact commit. But a similar commit made bots too campy in some occasions.

In my opinion we should avoid to have `default.bt` make bots campy at all. It would be better to make a nicer `camper.bt`, and give players the freedom to choose behaviors. This could add a strategic dimension to games with few players.
